substrate/frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+//! * The [`Config::TargetList`] sorting must be *always* kept up to date, even in the event of new
+//! nomination updates, nominator/validator slashes and rewards.
+//!
+//! Note that from the POV of this pallet, all events will result in one or multiple
+//! updates to the [`Config::VoterList`] and/or [`Config::TargetList`] state. If a update or set of
+//! updates require too much weight to process (e.g. at nominator's rewards payout or at nominator's
+//! slashes), the event emitter should handle that in some way (e.g. buffering events).
+//!
+//! ## Event emitter ordering and staking ledger state updates
+//!
+//! It is important to ensure that the events are emitted from staking (i.e. the calls into
+//! [`OnStakingUpdate`]) *after* the caller ensures that the state of the staking ledger is up to
+//! date, since the new state will be fetched and used to update the sorted lists accordingly.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+use frame_election_provider_support::SortedListProvider;
+use frame_support::traits::{Currency, Defensive};
+use sp_staking::{
+	currency_to_vote::CurrencyToVote, OnStakingUpdate, StakerStatus, StakingInterface,
+};
+use sp_std::collections::btree_map::BTreeMap;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+/// The balance type of this pallet.
+pub type BalanceOf<T> = <<T as Config>::Staking as StakingInterface>::Balance;
+/// The account ID of this pallet.
+pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+/// Represents a stake imbalance to be applied to a staker's score.
+#[derive(Copy, Clone, Debug)]
+pub enum StakeImbalance<Balance> {
+	// Represents the reduction of stake by `Balance`.
+	Negative(Balance),
+	// Represents the increase of stake by `Balance`.
+	Positive(Balance),
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use crate::*;
+	use frame_election_provider_support::VoteWeight;
+	use frame_support::pallet_prelude::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Currency: Currency<Self::AccountId, Balance = BalanceOf<Self>>;
+
+		/// The staking interface.
+		type Staking: StakingInterface<AccountId = Self::AccountId>;
+
+		/// Something that provides a best-effort sorted list of voters.
+		///
+		/// To keep the load off the chain as much as possible, changes made to the staked amount
+		/// via rewards and slashes are dropped and thus need to be manually updated through
+		/// extrinsics. In case of `bags-list`, this always means using `rebag` and `putInFrontOf`.
+		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
+
+		/// Something that provides a best-effort sorted list of targets.
+		///
+		/// Note that while at the time of nomination all targets are checked to be real
```

Comment:
Not sure if this comment applies at this point.

Reply:
Right, removed.